### ISSUE TITLE
Refactor payment gateway instance creation 

### ISSUE THREAD

**Issue by FangedParakeet** [Automattic/woocommerce-payments#5998]
### Description

The WCPay plugin will soon support four sibling payment gateways (legacy card element, legacy UPE payment element/"new checkout experience", split-UPE, new UPE/split-UPE with deferred intent confirmation). Maintaining this complex matrix of compatibility is leaving us with a [mammoth section of code](https://github.com/Automattic/woocommerce-payments/blob/poc/deferred-upe-intent-creation/includes/class-wc-payments.php#L445-L483) where these gateways are instantiated, on the basis of various enabled feature flags. 

There is a similiar murky melange of code involved in [registering the required payment gateway or gateways](https://github.com/Automattic/woocommerce-payments/blob/poc/deferred-upe-intent-creation/includes/class-wc-payments.php#L653-L697). These clusters of code are getting quite cluttered and on a cursory scan appears as if portions of repeated code might be collapsible. Once our [PoC/MVP new UPE implementation is merged](https://github.com/Automattic/woocommerce-payments/pull/5869), we should absolutely return and refactor these sections of code. 

One sensible suggestion is to engineer a new factory class that will handle this logic, but essentially we would like to extract this logic from within the `WC_Payments::init()` function and place it somewhere out of sight. These will keep our main `WC_Payments` class more clear and consistent as and when existing and future payment gateways are respectively removed or added. 

### Acceptance criteria

* Extract gateway branching logic from `WC_Payments` class.
* Do not affect any existing functionality.
* Add unit tests for refactored solution
* etc.

### Additional context

Blocked until #5869 has been completed.

**Comment by frosso** [comment]
Let's pretend like we worked really hard to refactor this, when in reality we're just naturally removing old code 👍
